synapse/handlers/message.py

```diff
@@ -752,20 +752,14 @@ async def _is_exempt_from_privacy_policy(
         if builder.type == EventTypes.Member:
             membership = builder.content.get("membership", None)
             if membership == Membership.JOIN:
-                return await self._is_server_notices_room(builder.room_id)
+                return await self.hs.get_room_member_handler().is_server_notice_room(
```

Comment:
This introduces a dependency `MessageHandler -> RoomMemberHandler`.

Does it make sense for handlers to interdepend?

(I expect this is no worse than we do elsewhere, so this is probably fine)

Reply:
This is one of those things that I don't know the answer to myself, actually. It seems objectively worse to have the redundancy though, but maybe I'll ask around...

Reply:
We try and avoid it where possible, as we need to make sure that `RoomMemberHandler` is initialised before we start the homeserver (as handlers may raise on config errors etc), and doing inline `hs.get_*` obviously means we can't necessarily guarantee that if all *other* usages also get inlined. 

In this particular case: it feels like `is_server_notices_room` could either be moved into `ServerNoticesManager` or into the data store?

Reply:
I never know what the distinction is meant to be between managers and handlers, but a note:

`ServerNoticesManager` calls into `RoomMemberHandler` (it takes a reference to it in `__init__` too).

If I move the function into there, we'll still have a circular dependency.

I didn't really realise that `get_*` was doing lazy initialisation.

I'll try putting it into the data store.